### PR TITLE
feat: Azbatch executor (#1953)

### PR DESCRIPTION
**Pull request by johanneskoester**
### Description

Finish remaining issues with the original implementation of the Azure Batch Executor originally tackled by @andreas-wilm in #533.

- simplified command line options
- batch configuration as a class with sensible defaults, or environment variable overrides.
- made use of existing AzBlob helper where possible.
- updates dockerfile with necessary deps
- container jakevc/snakemake:latest contains these changes for testing

Toyed with the idea of having more configuration from command line options, but this litters the codebase with redundant code (each command line option needs to be repeated a few times in __init__.py, scheduler.py, workflow.py, azure_batch.py). Env Variables seem appropriate with sensible defaults.

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### Example

Example invocation:

```bash
export AZ_BLOB_ACCOUNT_URL=*********
export AZ_BATCH_ACCOUNT_KEY=*********

snakemake \
    --jobs 3 \
    -rpf --verbose --default-remote-prefix snaketest \
    --use-conda \
    --default-remote-provider AzBlob \
    --envvars AZ_BLOB_ACCOUNT_URL \
    --az-batch \
    --container-image jakevc/snakemake \
    --az-batch-account-url **********
```

---------

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### CODE REVIEW

**Comment by johanneskoester** [comment]
@jakevc now the tests finally run, but there is an error. Can you have a look?

**Comment by jakevc** [comment]
Was able to reproduce the error, looking into it.

**Comment by jakevc** [comment]
This failure I'm having a hard time reproducing... as I am able to pass locally against my remote resources. @johanneskoester two questions
1. Is the storage account in the same region as the batch account ? (I think this is necessary)
2. Is the storage account traditional "blob" or  was "Azure Data Lake Storage gen2"  enabled? (I have not confirmed the effect of this on the use of the APIs, but I am passing with a traditional "blob" storage account.)

**Comment by vsmalladi** [comment]
@johanneskoester @jakevc 

Any luck?

**Comment by jakevc** [comment]
Okay new hypothesis, I think the SECRETS here might be using a storage account key, but only SAS token is supported the way this extension is designed.

**Comment by jakevc** [comment]
Ahh okay confirmed we are failing with 

` AssertionError: AZ_BLOB_CREDENTIAL does not match the SAS token pattern.`

**Comment by vsmalladi** [comment]
Awesome that is great news.

**Comment by johanneskoester** [comment]
@jakevc there are currently some permission problems in the Snakemake-org account. I am trying to fix them with MS, and then will change the token.

File: docs/executor_tutorial/azure_batch.rst [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+.. code:: console
+
+    # can set variables as appropriate
+    export accountname=snakebatch01
+    az batch account create --resource-group $resgroup --name $accountname --location $region
+
+
+The format of the batch account url is :code:`https://${accountname}.${region}.batch.azure.com`, which can be constructed manually from the output of the command :code:`az batch account list` or copied from the azure portal overview page of your batch account.
+
+
```
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:125]: Add example

export batch=$(az batch account show --name $accountname --resource-group $resgroup --query "accountEndpoint" --output tsv)
export batch_account_url=="https://$batch"

File: docs/executor_tutorial/azure_batch.rst [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+To run the test workflow, two primary environment variables need to be set local to the snakemake invocation.
+The azure batch account key, and the azure storage account url with an SAS key. More details about the AZ_BLOB_ACCOUNT_URL 
+are described in the section below. 
+
+.. code:: console
+
+     export AZ_BLOB_ACCOUNT_URL='${storage_account_url_with_sas}'
+     export AZ_BATCH_ACCOUNT_KEY='${az_batch_account_key}'
```
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:139]: `
 export AZ_BLOB_ACCOUNT_URL="${storage_account_url_with_sas}"
 export AZ_BATCH_ACCOUNT_KEY="${az_batch_account_key}"

`

File: docs/executor_tutorial/azure_batch.rst [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+Now you are ready to run the analysis:
+
+.. code:: console
+
+    # required env variables
+    export AZ_BLOB_PREFIX=snakemake-tutorial
+    export AZ_BATCH_ACCOUNT_URL='${batch_account_url}'
+    export AZ_BATCH_ACCOUNT_KEY='${az_batch_account_key}'
+    export AZ_BLOB_ACCOUNT_URL='${account_url_with_sas}'
```
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:259]: Doesn't work with single quotes on Ubuntu. Use double quotes. 
```
export AZ_BATCH_ACCOUNT_URL="${batch_account_url}"
export AZ_BATCH_ACCOUNT_KEY="${az_batch_account_key}"
export AZ_BLOB_ACCOUNT_URL="${account_url_with_sas}"
```

File: docs/executor_tutorial/azure_batch.rst [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+    conda create -c bioconda -c conda-forge -n snakemake snakemake
+
+Make sure that the ``azure-batch`` and ``azure-storage-blob`` modules are installed
+in this environment. Should they be missing, they can be installed with:
+
+.. code:: console
+
+   pip install azure-batch
+   pip install azure-storage-blob
```
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:49]: Also install 
```
pip install azure-mgmt-batch
pip install azure-identity

```

**Comment by vsmalladi** [comment]
Any way to unblock this?

**Comment by johanneskoester** [comment]
I think I made it finally. However, I think your docs (also the docs for the azure remote provider) should get some instructions on how to create these tokens, what permissions they should get, and what exactly to put into the variables (with examples). Because this whole azure interface is really complex ;-).

**Comment by vsmalladi** [comment]
> I think I made it finally. However, I think your docs (also the docs for the azure remote provider) should get some instructions on how to create these tokens, what permissions they should get, and what exactly to put into the variables (with examples). Because this whole azure interface is really complex ;-).

Can you point to what docs you are using and I will look to see how to update everthing.

**Comment by johanneskoester** [comment]
> > I think I made it finally. However, I think your docs (also the docs for the azure remote provider) should get some instructions on how to create these tokens, what permissions they should get, and what exactly to put into the variables (with examples). Because this whole azure interface is really complex ;-).
> 
> Can you point to what docs you are using and I will look to see how to update everthing.

I've merged it for now. I meant that probably the snakemake docs on azure should give some details how to obtain all those tokens. But it should be fine to do that in a separate PR if you like.

**Comment by johanneskoester** [comment]
Thanks a lot for all your work here! This was the last blocker for updating the paper. Hopefully I can finalize that soon, too.

**Pull request merged**